src/app/binding/BindingTable.cpp

```diff
@@ -0,0 +1,237 @@
+    mOnDeviceConnectionFailure(OnDeviceConnectionFailureFn, this)
+{}
+
+CHIP_ERROR BindingTable::Add(EndpointId sourceEndpoint, NodeId peerNodeId, EndpointId peerEndpointId, FabricIndex peerFabricIndex,
+                             ClusterId clusterId)
+{
+    BindingTableEntry * entry =
+        mBindingEntries.CreateObject(sourceEndpoint, peerNodeId, peerEndpointId, peerFabricIndex, clusterId);
+    if (entry == nullptr)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+    if (FindPeerDevice(entry->mPeerId) == nullptr)
+    {
+        DeviceProxyInitParams initParams = { &Server::GetInstance().GetSecureSessionManager(),
+                                             &Server::GetInstance().GetExchangeManager(),
+                                             &Server::GetInstance().GetSessionIDAllocator(),
+                                             GetFabricInfoFromIndex(peerFabricIndex) };
+        OperationalDeviceProxy * device  = mPeerDevices.CreateObject(initParams, entry->mPeerId);
+        if (device == nullptr)
+        {
+            mBindingEntries.ReleaseObject(entry);
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        return TryConnectDevice(device);
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BindingTable::Remove(EndpointId sourceEndpoint, NodeId peerNodeId, EndpointId peerEndpointId,
+                                FabricIndex peerFabricIndex, ClusterId clusterId)
+{
+    BindingTableEntry * removeEntry = nullptr;
+    int peerNumBindings             = 0;
+    PeerId peer                     = BuildPeerId(peerNodeId, peerFabricIndex);
+    mBindingEntries.ForEachActiveObject(
+        [&removeEntry, &peerNumBindings, sourceEndpoint, peer, peerEndpointId, clusterId](BindingTableEntry * entry) {
+            if (entry->mSourceEndpoint == sourceEndpoint && entry->mPeerId == peer && entry->mPeerEndpointId == peerEndpointId &&
+                entry->mClusterId == clusterId)
+            {
+                removeEntry = entry;
+            }
+            if (entry->mPeerId == peer)
+            {
+                peerNumBindings++;
+            }
+            return true;
+        });
+    if (removeEntry == nullptr)
+    {
+        return CHIP_ERROR_KEY_NOT_FOUND;
+    }
+    mBindingEntries.ReleaseObject(removeEntry);
+    // Last binding with the peer, close the connection.
+    if (peerNumBindings == 1)
+    {
+        OperationalDeviceProxy * device = FindPeerDevice(peer);
+        if (device)
+        {
+            mPeerDevices.ReleaseObject(device);
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BindingTable::TryConnectDevice(OperationalDeviceProxy * device)
+{
+    if (device == nullptr)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    if (device->IsConnected())
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    if (device->IsConnecting())
+    {
+        return CHIP_NO_ERROR;
+    }
+    Dnssd::Resolver::Instance().SetResolverDelegate(&mResolveDelegate);
```

Comment:
Isn't this clobbering any other resolver delegate that might have been there?

Reply:
Fortunately we're the only user of the resolver on end devices.

Reply:
What guarantees that?  This is a public API; the application code could call it if it wanted to, and nothing tells it not to, as far as I can tell.